### PR TITLE
exp run: throw error if new parameters are added by --set-param

### LINKED ISSUE #5477: exp run: validate params filename and parameter names
One thing I noticed when trying the changes from https://github.com/iterative/dvc/issues/5302 is that if the user specifies a parameter or params file that doesn't exist (i.e. typo's a param name) we just silently accept it and create the new file or new entry in the existing params file (this behavior existed before the `-S/--set-param` change as well).

If the specified parameter doesn't exist in the target stage/pipeline we should probably throw an error instead of silently ignoring it.

### PR DESCRIPTION
**Pull request by mattlbeck**
Throws an error if `dvc run exp --set-param` references a parameter key path
that is not in the target config file.

Resolves #5477 

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by mattlbeck** [comment]
This is now the minimum required to resolve the main topic of #5477 . I expect there should be some additional change to allow setting new parameters via `exp run` if the user explicitly wants to do this e.g. downgrade to a warning or use an additional flag. This is still up for discussion.

**Comment by mattlbeck** [comment]
Just noticed #6423 . I will have a think about other ways to implement the keypath set operation that doesn't use benedict.

File: tests/func/experiments/test_experiments.py [hunk 1]
```diff
@@ -399,9 +418,7 @@ def test_untracked(tmp_dir, scm, dvc, caplog, workspace):
 
     # copy.py is untracked
     with caplog.at_level(logging.ERROR):
-        results = dvc.experiments.run(
-            stage.addressing, params=["foo=2"], tmp_dir=True
-        )
+        results = dvc.experiments.run(stage.addressing, tmp_dir=True)
```
Comment (mattlbeck) [inline tests/func/experiments/test_experiments.py:421]: This needs careful review to make sure I am not invalidating this test.

I don't believe this call to exp run needs to be modifying the parameter as there appears to be no params file at the time of running (thus throws the new error unexpectedly) and the file is checked afterwards, as it is in the second exp run in the test. But there could be something to the test that I am not understanding.

File: dvc/repo/experiments/__init__.py [hunk 2]
```diff
@@ -350,6 +350,34 @@ def _pack_args(self, *args, **kwargs):
...
+        b_src = benedict(src)
+
+        return list(set(to_update.keys()) - set(b_src.keypaths(indexes=True)))
+
+    def _format_new_params_msg(self, new_params, config_path):
+        """Format an error message for when new parameters are identified"""
+        new_param_count = len(new_params)
+        additional_msg = "is"
+        if new_param_count > 1:
+            pluralise = "s are" if new_param_count > 2 else " is"
```
Comment (karajan1001) [inline dvc/repo/experiments/__init__.py:372]: I think here we have at least 2 params, it is always `are` here

File: dvc/repo/experiments/__init__.py [hunk 3]
```diff
@@ -350,6 +350,34 @@ def _pack_args(self, *args, **kwargs):
...
+            additional_msg = (
+                f"and {new_param_count-1} other " f"parameter{pluralise}"
```
Comment (karajan1001) [inline dvc/repo/experiments/__init__.py:374]: Maybe list all of the new_params are better? `param_a,param_b,param_c,...`

**Comment by karajan1001** [comment]
Any conclusions?

**Comment by mattlbeck** [comment]
I am torn between adding a new `--add-param` flag or just downgrading to a warning. Would love @dberenbaum's input befor proceeding.

**Comment by daavoo** [comment]
> Just noticed #6423 . I will have a think about other ways to implement the keypath set operation that doesn't use benedict.

Do you see this viable (implementing without using benedict)? As commented in https://github.com/iterative/dvc/pull/6484#discussion_r696278507 adding another place where benedict is used could be a drawback

**Comment by dberenbaum** [comment]
Thanks for taking this on yourself @mattlbeck! We haven't seen a great use case yet for using `--set-param` to add a new parameter, and I'm not in favor of adding more flags unless they add obvious value since this command is busy. If an error is more helpful than a warning for CI or greater visibility, then we can leave the usage as is and revisit if users find it overly strict (assuming the implementation questions like benedict usage get resolved of course).

**Comment by mattlbeck** [comment]
@dberenbaum sounds very sensible. I would personally prefer an error too - will leave this as it's current implementation and move on to address the earlier feedback.

**Comment by mattlbeck** [comment]
> > Just noticed #6423 . I will have a think about other ways to implement the keypath set operation that doesn't use benedict.
> 
> Do you see this viable (implementing without using benedict)? As commented in [#6484 (comment)](https://github.com/iterative/dvc/pull/6484#discussion_r696278507) adding another place where benedict is used could be a drawback

Commit 8d219dc4c4be678f8e12a57f766d282d13997443 adapts benedict's keypath parsing into the dvc codebase. I have taken the strategy of minimising the amount of code I copy from benedict, so instead of constructing a list of keypaths from the src dict I parse the incoming key paths and test them against the src dict. This may open up some corner cases that I haven't thought of yet...

**Comment by mattlbeck** [comment]
Does this change need a mention in the docs?

**Comment by dberenbaum** [comment]
> Does this change need a mention in the docs?

Yes, see the note in https://dvc.org/doc/user-guide/experiment-management/running-experiments#updating-experiment-parameters-on-the-fly. Thanks for checking!

File: dvc/repo/experiments/__init__.py [hunk 4]
```diff
@@ -362,6 +401,10 @@ def _update_params(self, params: dict):
             suffix = path.suffix.lower()
             modify_data = MODIFIERS[suffix]
             with modify_data(path, fs=self.repo.fs) as data:
+                new_params = self._get_new_params(data, params[params_fname])
+                if new_params:
+                    error_msg = self._format_new_params_msg(new_params, path)
+                    raise DvcException(error_msg)
```
Comment (daavoo) [inline dvc/repo/experiments/__init__.py:407]: Rasing a new exception created for this case (`ParamDoesNotExistError`?) or an existing one (`MissingParamsError`? from `dvc.dependency.param`) would be probably better than raising the generic `DvcException`.

File: dvc/repo/experiments/__init__.py [hunk 5]
```diff
@@ -350,6 +352,44 @@ def _pack_args(self, *args, **kwargs):
...
+                subdict = subdict[key]
+            except (KeyError, IndexError, TypeError) as e:
+                logger.debug(e)
+                return True
+        return False
+
+    def _format_new_params_msg(self, new_params, config_path):
+        """Format an error message for when new parameters are identified"""
+        new_param_count = len(new_params)
+        pluralise = "s were" if new_param_count > 1 else " was"
```
Comment (jorgeorpinel) [inline dvc/repo/experiments/__init__.py:386]:
```suggestion
        pluralise = "s are" if new_param_count > 1 else " is"
```

File: dvc/repo/experiments/__init__.py [hunk 6]
```diff
@@ -350,6 +352,44 @@ def _pack_args(self, *args, **kwargs):
...
+        param_list = ", ".join(new_params)
+        return (
+            f"{new_param_count} parameter{pluralise} missing "
+            f"from {config_path}: {param_list}"
```
Comment (jorgeorpinel) [inline dvc/repo/experiments/__init__.py:390]:
```suggestion
            f"from '{config_path}': {param_list}"
```

**Comment by mattlbeck** [comment]
@dberenbaum I think the PR is just awaiting a further review after having addressed the feedback

**Comment by skshetry** [comment]
@mattlbeck, let's just use `benedict` for now, we'll look for better alternatives later.

**Comment by skshetry** [comment]
Hi, @mattlbeck, let's use `benedict`. But you may need to import from `dvc.collections._benedict` like done here:

File: dvc/repo/experiments/__init__.py [hunk 7]
```diff
@@ -362,6 +387,10 @@ def _update_params(self, params: dict):
             suffix = path.suffix.lower()
             modify_data = MODIFIERS[suffix]
             with modify_data(path, fs=self.repo.fs) as data:
+                new_params = self._get_new_params(data, params[params_fname])
+                if new_params:
+                    error_msg = self._format_new_params_msg(new_params, path)
+                    raise MissingParamsError(error_msg)
```
Comment (skshetry) [inline dvc/repo/experiments/__init__.py:393]: @mattlbeck, could you please move this to `merge_params`? That way, you can just test it via unittest, though you may need to adjust existing ones. Although, we could extend `merge_params`'s API with `only_update=True|False`.

File: dvc/repo/experiments/__init__.py [hunk 8]
```diff
@@ -350,6 +351,30 @@ def _pack_args(self, *args, **kwargs):
...
+
+        src is a traditionally nested dict structure representing a parameter
+        config file. to_update is a flat dict with keypath keys representing
+        param names.
+        """
+        from dvc.utils._benedict import benedict
+
+        b_src = benedict(src)
+
+        return list(set(to_update.keys()) - set(b_src.keypaths(indexes=True)))
```
Comment (skshetry) [inline dvc/repo/experiments/__init__.py:366]: Does this work for `optimizer.lr=0.01` like nested-parameters?

File: tests/func/experiments/test_experiments.py (hunk 1 above)
Comment (pmrowla) [reply tests/func/experiments/test_experiments.py:421]: This test is checking for the behavior that happens when any important files are untracked (completely new file from git's perspective). In this case, with the new behavior we end up hitting the error for having a new parameter first (before we get to the relevant checks for the untracked files behavior).

It would be better to revert this change and also generate `params.yaml` as a git-tracked file from the start (so that the params existence is irrelevant for this test).

So these changes should be reverted, and line 409 should become:

```py
    tmp_dir.scm_gen("params.yaml", "foo: 1", commit="track params")
```

**Review by pmrowla — CHANGES_REQUESTED** [review]

File: dvc/repo/experiments/__init__.py (hunk 5 above)
Comment (jorgeorpinel) [reply dvc/repo/experiments/__init__.py]: Thanks
